pallets/registrar/src/lib.rs

```diff
@@ -159,37 +173,52 @@ pub mod pallet {
+        /// Tried to mark_valid_for_collating a ParaId that is not in PendingVerification
+        ParaIdNotInPendingVerification,
     }
 
     #[pallet::call]
     impl<T: Config> Pallet<T> {
         /// Register parachain
         #[pallet::call_index(0)]
-        #[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,2).ref_time())]
+        #[pallet::weight(10_000 + T::DbWeight::get().reads_writes(3,2).ref_time())]
         pub fn register(
             origin: OriginFor<T>,
             para_id: ParaId,
             genesis_data: ContainerChainGenesisData,
         ) -> DispatchResult {
-            T::RegistrarOrigin::ensure_origin(origin)?;
-            Self::schedule_parachain_change(|para_ids| {
-                // We don't want to add duplicate para ids, so we check whether the potential new
-                // para id is already present in the list. Because the list is always ordered, we can
-                // leverage the binary search which makes this check O(log n).
-                let result = match para_ids.binary_search(&para_id) {
-                    Ok(_) => Err(Error::<T>::ParaIdAlreadyRegistered.into()),
-                    Err(index) => {
-                        para_ids
-                            .try_insert(index, para_id)
-                            .map_err(|_e| Error::<T>::ParaIdListFull)?;
+            let account = ensure_signed(origin)?;
+            // Send deposit to reserve account
+            T::Currency::transfer(
+                &account,
+                &Self::account_id(),
+                T::DepositAmount::get(),
+                KeepAlive,
+            )?;
+
+            // Check if the para id is already registered
+            let pending_paras = <PendingParaIds<T>>::get();
+            let base_paras = pending_paras
+                .last()
+                .map(|&(_, ref paras)| paras.clone())
+                .unwrap_or_else(Self::registered_para_ids);
+            match base_paras.binary_search(&para_id) {
+                Ok(_) => return Err(Error::<T>::ParaIdAlreadyRegistered.into()),
+                Err(_) => (),
+            }
 
-                        Ok(())
-                    }
-                };
-                result
-            })?;
+            // Insert para id into PendingVerification, if it does not exist there
+            let mut pending_verification = PendingVerification::<T>::get();
+            match pending_verification.binary_search(&para_id) {
+                Ok(_) => return Err(Error::<T>::ParaIdAlreadyRegistered.into()),
+                Err(index) => {
+                    pending_verification
+                        .try_insert(index, para_id)
+                        .map_err(|_e| Error::<T>::ParaIdListFull)?;
+                }
+            }
 
-            // TODO: while the registration takes place on the next session, the genesis data
+            // While the registration takes place on the next session, the genesis data
```

Comment:
This comment is a bit outdated, registration happens on the next session after someone else calls `mark_valid_for_collating`, I will reword it.